### PR TITLE
KEYCLOAK-19273 Return correct status when user is disabled temporarily

### LINKED ISSUE #11726: Conflicting data returned for /users/id and /users endpoints when user is temporarily locked
### Describe the bug

The endpoints `/users/id` and `/users` return different results for the `enabled` field for the same user when the user is temporarily locked.

### Version

16.0.0, 15.0.1

### Expected behavior

Both the endpoints should return the same data for the same user.

### Actual behavior

The user details via the `/users?username=user` endpoint wrongly returns `enabled: true`

### How to Reproduce?

1. Enabled brute force detection
2. Create user
3. Repeatedly login with user with wrong password till it gets locked

Get the user details via the `/users?username=user` endpoint. **Wrongly returns enabled: true**

```console
$ curl -k -X GET -H "Authorization: Bearer $TOKEN" -H "Content-Type: application/json"     $KC_URL/auth/admin/realms/master/users?username=user

[
  {
    "id": "21daeda5-86bc-406d-a687-fa4231bc6b54",
    "createdTimestamp": 1631277330288,
    "username": "user",
    "enabled": true,
    "totp": false,
    "emailVerified": false,
    "disableableCredentialTypes": [],
    "requiredActions": [],
    "notBefore": 0,
    "access": {
      "manageGroupMembership": true,
      "view": true,
      "mapRoles": true,
      "impersonate": true,
      "manage": true
    }
  }
]
```

Get the user details via the `/users/id` endpoint. **Correctly returns enabled: false**

```console
$ curl -k -X GET -H "Authorization: Bearer $TOKEN" -H "Content-Type: application/json"     $KC_URL/auth/admin/realms/master/users/21daeda5-86bc-406d-a687-fa4231bc6b54

{
  "id": "21daeda5-86bc-406d-a687-fa4231bc6b54",
  "createdTimestamp": 1631277330288,
  "username": "user",
  "enabled": false,
  "totp": false,
  "emailVerified": false,
  "disableableCredentialTypes": [],
  "requiredActions": [],
  "notBefore": 0,
  "access": {
    "manageGroupMembership": true,
    "view": true,
    "mapRoles": true,
    "impersonate": true,
    "manage": true
  }
}
```

### Anything else?

This issue was migrated from https://issues.redhat.com/browse/KEYCLOAK-19273.  It was originally created 2021/09/10 by @bhorkarg.

### PR DESCRIPTION
**Pull request by tsaarni**
BruteForceProtector was not checked for user being temporary disabled when
listing users using /{realm}/users endpoint.

Signed-off-by: Tero Saarni <tero.saarni@est.tech>

Fixes #11726

### CODE REVIEW

**Comment by tsaarni** [comment]
@hmlnarik, @pedroigor  Could you look into this pull request, related to https://issues.redhat.com/browse/KEYCLOAK-19273. It would be great to fix this. Thank you!

**Comment by pedroigor** [comment]
@tsaarni Thanks. Would you be able to add some tests to this class [1]?

[1] https://github.com/keycloak/keycloak/blob/c9e1e00b95d9f9f7a6f11c7e0c341781ed234c7b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/BruteForceTest.java

**Comment by tsaarni** [comment]
@pedroigor I've added a simple test case. If run with main branch (without the fix) it would fail, but it succeeds with this fix.

File: services/src/main/java/org/keycloak/services/resources/admin/UsersResource.java [hunk 1]
```diff
@@ -453,6 +454,9 @@ public UserProfileResource userProfile() {
                             ? ModelToRepresentation.toBriefRepresentation(user)
                             : ModelToRepresentation.toRepresentation(session, realm, user);
                     userRep.setAccess(usersEvaluator.getAccess(user));
+                    if (session.getProvider(BruteForceProtector.class).isTemporarilyDisabled(session, realm, user)) {
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/services/resources/admin/UsersResource.java:457]: One thing that it should be simple to include and help to not impact runtime performance when brute force is disabled is to make sure we only run this code if the realm has bruteforce enabled. 

You can infer that by looking at `RealmModel.isBruteForceProtected`.

**Review by pedroigor — CHANGES_REQUESTED** [review]

File: services/src/main/java/org/keycloak/services/resources/admin/UsersResource.java (hunk 1 above)
Comment (tsaarni) [reply services/src/main/java/org/keycloak/services/resources/admin/UsersResource.java]: I've added a check for this.

**Comment by tsaarni** [comment]
@pedroigor Is it possible for you to restart the test run? The "Model Tests" job might have got stuck, it has been running over 5 hours.

**Comment by tsaarni** [comment]
@pedroigor Kind reminder. Looking forward to get this merged! :)

**Comment by stianst** [comment]
@pedroigor @hmlnarik @mposolda although technically correct I don't think this can be merged in its current form as it will have a fairly big performance impact when listing users. WDYT?

**Comment by pedroigor** [comment]
@stianst Yes, it adds an additional round-trip to cache. But I don't see a big performance impact.

And like you said, the new behavior is indeed the correct one.

**Comment by tsaarni** [comment]
Kind reminder again :) It would be great to get this bug fixed. What would be the way forward?

**Review by pedroigor — APPROVED** [review]
@stianst @mposolda In general, LGTM.

If you agree we can squash the commits prior to merging.

**Review by stianst — CHANGES_REQUESTED** [review]
I'm not in favour of adding this as it will have a significant impact on performance when fetching a larger number of users. Especially if there's not a cache in place.

**Comment by tsaarni** [comment]
> Especially if there's not a cache in place.

If I understand correctly, it is known that this fix will not perform adequately and that it will not be accepted even if benchmarking it together with the cache?

Is there any alternative approach that I could take to fix this bug?  I would appreciate some guidance, even if it is that I should stop pursuing for a fix.

**Comment by mposolda** [comment]
AFAIK the fix adds calling of `UserLoginFailureProvider.getUserLoginFailure(RealmModel realm, String userId)` separately for each user, which is going to be returned from the search.

I am not sure how it is done in the new store and what is the performance impact of this here? In the old store, there is separate request to infinispan `loginFailures` cache, which is `distributable` cache by default (in cluster environment) and hence there is potential impact as user entry may not be available in the current "local" node.

I wonder if it makes sense to add some method to `UserLoginFailureProvider` like:
```
Stream<UserLoginFailureModel> getUserLoginFailures(RealmModel realm, Stream<userIds> users);
```
but this will need to be discussed/agreed with new store team if it is something, which would makes sense. I think that for old store (infinispan based impl), it will probably help with the performance and reduce the number of cluster calls if it is done in "bulk" rather than in separate calls. But no clue about new store.

@hmlnarik @martin-kanis WDYT?

**Comment by pedroigor** [comment]
I think that saying it has a big impact on performance is subjective. I understand it might cause state transfer between nodes but it can be acceptable and managed depending on your use case.

The User API also allows pagination/filter and you don't really want to return all users from there in any case.

**Comment by stianst** [comment]
@pedroigor huh? You're going from N queries to the database to N+1, so if page size is 100, that's 101 queries to the DB compared to 1. Makes no sense that you can argue that won't have a significant impact on performance.

**Comment by pedroigor** [comment]
> @pedroigor huh? You're going from 1 query to the database to N+1, so if page size is 100, that's 101 queries to the DB compared to 1. Makes no sense that you can argue that won't have a significant impact on performance.

Isn't login failures stored in infinispan?

**Comment by pedroigor** [comment]
Indicating the user is disabled from this API gives the wrong message because the user is not really disabled but temporarily disabled as per brute force provider rules.

Perhaps we don't need this PR at all because if you want to query if the user is temporarily disabled you can reach out  `/admin/realms/${realm}/attack-detection/brute-force/users/:userId`.

**Comment by tsaarni** [comment]
> Indicating the user is disabled from this API gives the wrong message because the user is not really disabled but temporarily disabled as per brute force provider rules.

If defining `enabled` in that way, we already have `/{realm}/users/{id}` which currently checks BruteForceProtector #11726.

**Comment by stianst** [comment]
> > @pedroigor huh? You're going from 1 query to the database to N+1, so if page size is 100, that's 101 queries to the DB compared to 1. Makes no sense that you can argue that won't have a significant impact on performance.
> 
> Isn't login failures stored in infinispan?

In the legacy store it is stored in Infinispan, but it's a distributed cache right, so if there are more than 1 node in the cluster that would result in a fair bit of external requests for each user listing.

In the new store these could very well be stored in a relational DB.

Main point here though is that none or only some of the login failures will be locally in most cases, resulting in this having a potentially large impact on the performance of this endpoint.

**Comment by stianst** [comment]
> > Indicating the user is disabled from this API gives the wrong message because the user is not really disabled but temporarily disabled as per brute force provider rules.
> 
> If defining `enabled` in that way, we already have `/{realm}/users/{id}` which currently checks BruteForceProtector #11726.

That's somewhat not correct either, and it's a lot less of an issue compared to what's being suggested here, as it's looking up the login failures for a single user, not for a list of users.

**Comment by stianst** [comment]
I don't think there is a sensible way to get this PR into a mergable state, and also disagree to some extent that it should be included, as @pedroigor points out there's a difference to temporarily disabled by brute force protection, and disabled fully. I think rather we should consider reworking the fetch user endpoint to not include temporarily disabled by brute force when returning if the user is disabled or not.

**Comment by tsaarni** [comment]
Fair enough. My intention was of course not to propose anything outrageous! Sorry for not realizing the implications.

As a summary (in my words):
- Due to performance reasons, it is not possible for `/{realm}/users` to include information about temporarily disabled users
- Endpoint `{realm}/attack-detection/brute-force/users/:userId` can be used to check if enable user is temporarily disabled
- The user `enable: true/false` is to be defined NOT to include information about user being temporarily disabled 

Thanks for looking into this and getting it into conclusion. I has been a long while.

**Comment by tsaarni** [comment]
Closing as not planned.

**Comment by pedroigor** [comment]
@tsaarni Thanks for your time on this.

**Comment by mposolda** [comment]
> I think rather we should consider reworking the fetch user endpoint to not include temporarily disabled by brute force when returning if the user is disabled or not.

This will be nice, but it is backwards incompatible change with potential security implications IMO. It is possible that people already rely on the fact that single-user endpoint `/{realm}/users/{id}` returns `enabled` with value `false` in case that user is temporarily disabled by brute force...

Hopefully we have versioned API as changing things returned from the REST API can be a pain due the backwards compatibility...:-/

**Pull request closed without merging**